File: Leetcode/2197 Replace Non-Coprime Numbers in Array.py
class Solution:
    def replaceNonCoprimes(self, nums):
        stack = []
        for n in nums:
            while (g := gcd(n, stack[-1] if stack else 1)) != 1:
                n *= stack.pop() // g
            stack.append(n)
        return stack

def gcd(a, b):
    return a if b == 0 else gcd(b, a % b)  

File: Leetcode/2197 Replace Non-Coprime Numbers in Array.ts
function replaceNonCoprimes(nums: number[]): number[] {
    const stack: number[] = [];
    for (const num of nums) {
        let n = num;
        while (true) {
            const g = gcd(stack.length === 0? 1 :  stack[stack.length-1], n);
            if (g === 1){
                break;
            }
            n *= (stack.pop() as number) / g;
        }    
        stack.push(n);
    }
    return stack;
}

function gcd(a: number, b: number): number {
    return b === 0 ? a : gcd(b, a % b);
}
